### ISSUE TITLE
Japanese Layout -- 106/109 key layout missing keys

### ISSUE THREAD

**Issue by AbstractUmbra** [Dygmalab/Bazecor#151]
**Describe the bug**
With my Dygma and Bazecor on Windows:

I have Windows set to Japanese input, 106/109 key: https://images.abstractumbra.xyz/NrCe6MVexilMZA.png
When set to this, Bazecor layout has some missing keys or irregularities.

I have no access to backslash, pipe, or the Yen key, typically the key to the left of つ/z. I currently have this mapped to "ISO <>" in Bazecor 0.2.6. 
The key does nothing at all, unless I switch Windows back to the 101/102 layout: https://images.abstractumbra.xyz/ebMuEbBZEUDpcQ.png but in that event, none of the punctuation or symbol keys work as intended nor as they are printed on my keys (Japanese).

Summary:
ろ key should give backslash or underscore with shift modifier, currently it switches between alphanumeric and kana input (which seems like a separate issue?).
¥ key should give pipe or underscore with shift modifier, currently it does nothing.

**To Reproduce**
0. (??) Set Windows input to 日本語/Japanese input, 106/109 key layout as in earlier screenshot.
1. Set Bazecor to Japanese layout: https://images.abstractumbra.xyz/dfgre1aL638fjQ.png
2. Attempt to use these keys: https://images.abstractumbra.xyz/dqKHx3ymeUkgQ.png
3. Experience error.

**Expected behavior**
I expect the keys to work as intended on a fully fledged Japanese keyboard, of which I own and used to test.

**Screenshots**
Screenshots added as necessary, no logs needed as far as I am aware.

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Bazecor Version: 0.2.6 (latest)

**Additional context**
N/A

**Comment by AlexDygma** [comment]
We have to take a look into the Japanese layout, something is clearly wrong.

It will be very useful if you could tell us exactly what behavior you are expecting from the keys at your disposal so we can make it work exactly that way.

**Comment by AbstractUmbra** [comment]
Hello! 
Thanks for the reply. Here are the issues and their expected outcomes:

I will use a keyboard tester to map the keycodes and type on an actual Japanese keyboard vs what Bazecor has mapped.

## Hankaku key
[Key Location](https://images.abstractumbra.xyz/n0U7amj2uwEtWA.png)
This key does not appear to be within Bazecor at all. This is used to switch between input types (DirectInput/Hiragana).
| Key Code |
| ---------- |
| [Standard Board](https://images.abstractumbra.xyz/mW2r0fYOEQptTQ.png) |
| Bazecor |

However the Ro/ろ key has this behaviour and same key code (243).

## Ro/ろ key
[Key Location](https://images.abstractumbra.xyz/YRt5qCIFDP9nkw.png)
This key should give a backslash (\) or underscore (_) with Shift Modifier, on my Raise it acts as the Hankaku key (above).
| Key Code |
| ---------- |
| [Standard Board](https://images.abstractumbra.xyz/MbjYKY6VX2NfFA.png) |
| [Bazecor](https://images.abstractumbra.xyz/gmMsV5PVq7hyhA.png) |

## Yen/Pipe/Backslash key
[Key Location](https://images.abstractumbra.xyz/DzuJLZd0PQSnOg.png)
This key should give a backslash (\) or pipe (|) with Shift Modifier, on my Raise it does nothing. I believe it is mapped in Bazecor as the "ISO <>" [key](https://images.abstractumbra.xyz/4sG2lbUerOfnRw.png).
| Key Code |
| ---------- |
| [Standard Board](https://images.abstractumbra.xyz/qdUTZ0m_08JgWA.png) |
| [Bazecor](https://images.abstractumbra.xyz/bMmknyBfJzfg9Q.png) |

## "Hiragana" key
[Key Location](https://images.abstractumbra.xyz/LVSy3TptQ3C5w.png)
This key should switch your input type to Hiragana. This does not appear to be present on Bazecor.
| Key Code |
| ---------- |
| [Standard Board](https://images.abstractumbra.xyz/Q_7ohyz_5Q0lOg.png) |
| Bazecor |

## Convert key
[Key Location](https://images.abstractumbra.xyz/UhJEqqtKA2zgg.png)
This key should allow you to 'convert' already entered text after highlighting due to accidentally entered Kanji.
| Key Code |
| ---------- |
| [Standard Board](https://images.abstractumbra.xyz/kJAUNf9kH64vhw.png) |
| Bazecor |

## NoConvert key
[Key Location](https://images.abstractumbra.xyz/JysAmmIkRFTi4Q.png)
At the moment I am unsure what this key does, I will update when I find out.
| Key Code |
| ---------- |
| [Standard Board](https://images.abstractumbra.xyz/dOzWxqrYyjEzSQ.png) |
| Bazecor |

## ContextMenu key
[Key Location](https://images.abstractumbra.xyz/MnbJ5ZmcrRve2w.png)
This key emulates a right click action and opens a Context Menu in your active window (not local to Japanese layout). I believe it to be currently missing from Bazecor.
| Key Code |
| ---------- |
| [Standard Board](https://images.abstractumbra.xyz/NhRfmc_U5O8yRQ.png) |
| Bazecor |

### Notes

The key tester I used is here: <https://codepen.io/chriscoyier/pen/mPgoYJ>

**Comment by AbstractUmbra** [comment]
Sorry for bumping like this - I am just wondering if there is possibly any update?

**Comment by AlexDygma** [comment]
Hey @AbstractUmbra !

We have been working mainly on the freezing issues and the new bazecor this past months so i had no time to close issues like this one 😞 

But this wont last! we are nearing the beta of the new bazecor and one of our priorities is to clean up and add support for many other keycodes, we will be updating this next month, so stay tuned! (or this month, as today is already April, 1) 😉

**Comment by AbstractUmbra** [comment]
Hello! Sorry for bumping again. Just wondering if there are any more updates or potentially a fix/workaround on this for now?
It is a tad tedious to have to copy and paste certain frequently needed characters when I need them. :-)

**Comment by GazHank** [comment]
I think that the main issue is that we need to add in the extrakey codes found on the [106 keyboard](https://www.w3.org/TR/uievents-code/#keyboard-106) as supplemental codes that we can send (the user can then map these however they like, perhaps to the thumb keys or a separate layer - I know everyone is looking forward to tap dancing and all of the interesting options it will provide :-)

The missing keys being:

- "IntlRo" (labelled \ろ)
- "KanaMode" (labelled カタカナ/ひらがな/ローマ字 katakana/hiragana/romaji)
- "IntlYen" (labelled ¥)
- "Convert" (labelled 変換 henkan) 
- "NonConvert" (labelled 無変換 muhenkan)

These use keycodes 0x87 - 0x8B (135 - 139)

The key we have labelled as "ISO <>" (IntlBackslash, code 0x64 or 100) is a little different as it isn't present on a Japanese 106 keyboard, so that probably explains why it doesn't seem to cause anything to happen when Windows is treating the keyboard as a 106 keyboard.

On the other hand the context menu function shouldn't need a code change since it's just Shift+ F10, so we can map that wherever we like using the current version of Bazecor.

@AbstractUmbra would it be correct to group these keys under Punctionation and Special Letters (per below), or should I move them to another group?

![image](https://user-images.githubusercontent.com/64541249/120077229-43852900-c0a1-11eb-88e9-166567b6e275.png)

(I think we need to use  "Hirigana" (or "Kana Mode") for the button text, since the verbose カタカナ/ひらがな/ローマ字 is a little too long, and even with spaces the button text doesn't wordwrap for this)

![image](https://user-images.githubusercontent.com/64541249/120077708-9f50b180-c0a3-11eb-8c9d-550e84c43707.png)


I did wonder about adding the full set of International Codes 1 - 10 into the base code (i.e. English) just in case people want to use them with other keyboard layouts (such as Korean 103 or Brazilian 104).  However I'm currently thinking that we should only add these codes into the Japanese code for now, since adding them into the English language (i.e. base) code could cause confusion (they just won't do anything unless you have your keyboard set to 106 or one of the other variants).

**Comment by AbstractUmbra** [comment]
@GazHank thank you for this!

I think these would all map under Punctuation and Special Letters just fine.
Sorry about the Context Menu key. I completely forgot Shift + F10 achieved the same. Nice catch!

Everything else you mentioned makes perfect sense I believe! 

Thanks again!

**Comment by GazHank** [comment]
@AbstractUmbra no problem, I wasn't aware of the context key mapping before today.  I'm old enough to remember when Microsoft made a big deal about new keyboards being release with these extra keys; I had no idea they had just hardcoded it to Shift+F10 😆 

Hopefully the proposed change will fix the issue for you, but if not then let me know and I'll take another look

Cheers

Gaz

**Comment by AbstractUmbra** [comment]
Hey @GazHank - finally got the nerve to build a dev version of Bazecor with the changes in your PR.

Works perfect on my end! Thank you very much for this.

**Comment by GazHank** [comment]
Hi @AbstractUmbra does that mean that you think the symbols for keycodes 55 and 136 just need to be flipped?

So:
```
  {
    code: 55,
    labels: {
      primary: "る"
    }
  },
  {
    code: 136,
    labels: {
      primary: "Hiragana",
      verbose: "カタカナ/ひらがな/ローマ字"
    },
    newGroupName: "Punctuation"
  },
```

become:
  ```
{
    code: 55,
    labels: {
      primary: "Hiragana",
      verbose: "カタカナ/ひらがな/ローマ字"
    }
  },
  {
    code: 136,
    labels: {
      primary: "る"
    },
    newGroupName: "Punctuation"
  },
```

One source of the code info (in addition to the previous link shared) is the [HID codes pages (85 - 89)](https://www.usb.org/sites/default/files/hut1_22.pdf)  I should highlight that the hid codes are all written as Hex in the file, so 87 becomes 135.  As you can see there is a little note against code 55 (hex 37) that "2Typically remapped for other languages in the host system", so that might explain what you are seeing.

Let me know if you think switching the codes is needed and I'll update the pull request 

Cheers

Gaz

**Comment by AbstractUmbra** [comment]
Hey @GazHank thanks again for the reply.

Code 53 in `japanese.js` belongs to ろ as well as your new entry that shows as `\ろ` is code 135. Unsure if this is intentional.
But it seems 53 needs to be the `半角/全角 漢字` key (half/full width kanji), not ろ.

My diff is now:

```diff
--- a/src/api/keymap/languages/japanese/japanese.js
+++ b/src/api/keymap/languages/japanese/japanese.js
@@ -23,7 +23,7 @@ const japaneseLetters = [
   {
     code: 53,
     labels: {
-      primary: "ろ"
+      primary: "半角/全角 漢字"
     }
   },
   {
@@ -300,6 +300,42 @@ const japaneseLetters = [
     labels: {
       primary: "め"
     }
+  },
+  {
+    code: 135,
+    labels: {
+      primary: "ろ"
+    },
+    newGroupName: "Punctuation"
+  },
+  {
+    code: 136,
+    labels: {
+      primary: "Hiragana",
+      verbose: "カタカナ/ひらがな/ローマ字"
+    },
+    newGroupName: "Punctuation"
+  },
+  {
+    code: 137,
+    labels: {
+      primary: "¥"
+    },
+    newGroupName: "Punctuation"
+  },
+  {
+    code: 138,
+    labels: {
+      primary: "変換"
+    },
+    newGroupName: "Punctuation"
+  },
+  {
+    code: 139,
+    labels: {
+      primary: "無変換"
+    },
+    newGroupName: "Punctuation"
   }
 ];
 ```

and I think this works perfectly.

What do you think?

**Comment by GazHank** [comment]
Ah, I think that makes sense, so code 53 is the "width" character shown under the Escape key in the [W3 doc for the 106 keyboard](https://www.w3.org/TR/uievents-code/#keyboard-106), I had wondered why we had a Ro key and an international Ro Key...

I guess I'm quite lucky with the UK keyboard layout on the Raise as the key between Escape and Tab is the only one I really miss; it must be a little more challenging with 6 or so missing keys (thank goodness for layers).

In that case shall we make 53 primary: "半角" (or perhaps "Width") and verbose: "半角/全角 漢字" - as I'm assuming the full japanese characters are too large to fit the keys on the screen (as per Hiragana / "カタカナ/ひらがな/ローマ字"), or is there an alternative label you think we should use for the key?

**Comment by AbstractUmbra** [comment]
I'm not entirely certain on those choices. I believe the fully verbose looks okay (since we have "caps lock" and others), so perhaps fully verbose here would be acceptable?

This looks okay to me, but I am not an aesthetic person so I prefer explicitness!
![electron_2021-06-11_14-28-47](https://user-images.githubusercontent.com/16031716/121693912-7b3c9980-cac1-11eb-862e-23585d5fc1f8.png)

**Comment by GazHank** [comment]
Cool, I've updated the Pull request https://github.com/Dygmalab/Bazecor/pull/231 I just need to sign off the change once I get back to my main machine.

ps - love the colour scheme :-D

**Comment by AbstractUmbra** [comment]
This was fixed with the merging of #231. Closing it now, thank you!
